### DISCUSSION TITLE
Duplicate queries after enabling insights middleware

### DISCUSSION THREAD

**Discussion by giilby** [algolia/instantsearch#5036] [General]
Per https://www.algolia.com/doc/api-reference/widgets/insights/js/, I've configured the insights middleware in order to track some basic usage and conversion events. However, after doing so, I noticed duplicate queries happening on page load. The first is typically for an anonymous user (e.g. `userToken: anonymous-957a8f35-8aca-41e4-b397-30f87ca448dd`), and the second is for a logged-in user (e.g. `userToken: 12345`). Is this expected behavior? I was a bit surprised by it, as I assumed `instantsearch.js` would continue to only execute a single query on page load. If it's not expected, I'm happy to file a bug report...just thought I'd ask the question first.

**Comment by Haroenv** [comment]
This would indeed be classed as a bug, however I can't reproduce it, the only query I see happening is the one one set with an explicit user token: https://codesandbox.io/s/clever-agnesi-5brum9?file=/index.html:1255-1300

Do you have a sandbox of how your setup is done?

**Reply by giilby** [reply]
Well, I tried real hard to reproduce this issue in a sandbox, but was unsuccessful. Notably, the issue only occurs intermittently in our production application, so it may be a timing issue of some sort. Here are additional details that might be helpful:

We're setting `useCookie: true` in `aa( 'init'`; we also set `userToken` (when it's available) in that same `init` call, rather than using `setUserToken`.

When two queries occur on page load, here are the stack traces for each (our application code is `asset-search.min.js`, and we've implemented a custom `searchFunction`):

First query:
```
searchFunction (asset-search.min.js?ver=1.0.0:formatted:85)
t.search (InstantSearch.ts:474)
(anonymous) (InstantSearch.ts:564)
(anonymous) (defer.ts:26)
Promise.then (async)
e (defer.ts:18)
value (InstantSearch.ts:519)
(anonymous) (asset-search.min.js?ver=1.0.0:formatted:445) - this is `search.start()`
(anonymous) (asset-search.min.js?ver=1.0.0:formatted:723)
```

Second query:
```
searchFunction (asset-search.min.js?ver=1.0.0:formatted:85)
t.search (InstantSearch.ts:474)
(anonymous) (InstantSearch.ts:564)
(anonymous) (defer.ts:26)
Promise.then (async)
e (defer.ts:18)
addWidgets (index.ts:344)
value (InstantSearch.ts:375)
e (createInsightsMiddleware.ts:121)
(anonymous) (search-insights@2.2.1:1)
T.i (search-insights@2.2.1:1)
(anonymous) (search-insights@2.2.1:1)
T (search-insights@2.2.1:1)
m (search-insights@2.2.1:1)
(anonymous) (search-insights@2.2.1:1)
(anonymous) (search-insights@2.2.1:1)
(anonymous) (search-insights@2.2.1:1)
```

So, it appears that whenever the second query happens, it's being triggered by `search-insights@2.2.1`. Please let me know if there are any other details I can provide; I'll also try a few other things in the sandbox to see if I can replicate it there.

Thanks for looking into this!

**Reply by giilby** [reply]
After updating to the latest versions of `instantsearch.js` and `algoliasearch` today, we were still experiencing this behavior. However, I did find this article: https://support.algolia.com/hc/en-us/articles/12812765253009-Why-do-I-see-a-double-query-after-adding-insights-to-my-site. Since we're bundling with Webpack and including our script at the bottom of the page, the suggested solution isn't very straightforward for us. Thus, to sidestep the issue, I configured `insights` like this (for now):

```
insights: {
  insightsInitParams: {
    useCookie: true
  }
}
```

By omitting `authenticatedUserToken` (which we were previously setting), I no longer see duplicate queries on page load.
